junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestableFactory.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.engine.junit5.descriptor;
+
+import static org.junit.gen5.commons.util.ReflectionUtils.loadClass;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.gen5.commons.util.Preconditions;
+import org.junit.gen5.commons.util.ReflectionUtils;
+
+public class JUnit5TestableFactory {
+
+	private static final String SEPARATORS = ":$#";
+
+	public JUnit5Testable fromUniqueId(String uniqueId, String engineId) {
+		Preconditions.notEmpty(uniqueId, "uniqueId must not be empty");
+		List<String> parts = split(uniqueId);
+		Preconditions.condition(parts.remove(0).equals(engineId), "uniqueId must start with engineId");
```

Comment:
Just to make my point: The Launcher should not influence the resolution of the unique id for an engine at all. How would it guarantee, that subsequent test runs will always have the same ids?

One example: Let's say we have two engines that have a conflicting local/engine-specific unique id (other cases dont lead to problems anyways). What could a Launcher do to make them unique? The only feasible solution is to add the engine id in front of the local/engine-specific id. If there is still a conflict, the Launcher must give up and throw an exception telling that the engines are not compatible. Why?

If the Launcher influences the unique id by some kind of "magic" due to the given engines on the classpath, the returned full unique id would not be stable. It might be, that a subsequent call to the Launcher has only one engine on the classpath leading to another mapping/building of the full unique id.

So I definitely go for the solution, that the engine must guarantee the globally uniqueness of its descriptors, because it just doesnt make sense to move this responsibility somewhere else. The launcher only has to prove, that the ids are unique, and if it detects a conflict, it must throw some kind of exception complaining about the incompatibility of the engines.

But yes, we (you and me) have different ideas about this. So we both made our points, let's have someone else to decide on it...